keps/sig-api-machinery/20190325-unions.md

```diff
@@ -0,0 +1,201 @@
+to allow conflict detection when a union field is flipped.
+
+The value of the discriminator is going to be set automatically by the apiserver
+when a new field is changed in the union, and will be interpreted as an
+intention to clear all the other fields when its value is changed. See section
+below.
+
+### Normalization process
+
+The normalization process (which runs before validating that the union property
+is verified), works as follows:
+- If there is a discriminator, and its value has changed, clear all fields but
+  the one specified by the discriminator,
+- If there is no discriminator, or if its value hasn't changed,
+  - if there is exactly one field set, set the discriminator (if exists) to that
+    value. Otherwise
+  - compare the fields set before and after. If there is exactly one field
+    added, set the discriminator (if present) to that value, and remove all
+    other fields. if more than one field has been added, leave the process so
+    that validation will fail.
+
+### Backward compatibilities properties
+
+This normalization process has a few nice properties, especially for dumb
+clients, when it comes to backward compatibility:
+- A dumb client that doesn't know which fields belong to the union can just set
+  a new field and get all the others cleared automatically
+- A dumb client that doesn't know about the discriminator is going to change a
+  field, leave the discriminator as it is, and should still expect the fields to
+  be cleared accordingly
+- A dumb client that knows about the discriminator can change the discriminator
+  without knowing which fields to clear, they will get cleared automatically
+
+
+### Validation
+
+Objects have to be validated AFTER the normalization process, which is going to
+leave multiple fields of the union if it can't normalize. As discussed in
+drawbacks below, it can also be useful to validate apply requests before
+applying them.
+
+## Drawbacks
+
+Since the proposal only normalizes the object after the patch has been applied,
+it is hard to fail if the patch is invalid. There are scenarios where the patch
+is invalid but it results in an unpredictable object:
+
+Object has a discriminator named `discriminator` for a union of fields `a` and `b`.
+
+Object currently looks like:
+
+```
+discriminator: a
+a: 1
+```
+
+A clearly wrong patch is sent:
+
+```
+discriminator: b
+a: 1
+```
+
+The merge object looks like this:
+
+```
+discriminator: b
+a: 1
+```
+
+Since the discriminator has changed, `a` is cleared by normalization, which is (or
+isn't) what the client intended to do.
+
+Also, given the same patch, if the object was:
+
+```
+discriminator: b
+b: 1
+```
+
+the merged object would look like:
+```
+discriminator: b
+a: 1
+b: 1
+```
+
+Since the discriminator hasn't changed, but `a` has been added, the discriminator
+will flip to `a`, and `b` will be dropped.
+
+Validating patches is not a problem that we want to tackle now, but we can
```

Comment:
Maybe label this section "future work"? I don't think we can leave this problem in forever.